compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -3869,28 +3876,71 @@ public void visit(BLangOnFailClause onFailClause, AnalyzerData data) {
         // Create a new block environment for the on-fail node.
         SymbolEnv onFailEnv = SymbolEnv.createBlockEnv(onFailClause.body, data.env);
         VariableDefinitionNode onFailVarDefNode = onFailClause.variableDefinitionNode;
+        Stack<LinkedHashSet<BType>> onFailErrTypes = data.commonAnalyzerData.errorTypes;
+
         if (onFailVarDefNode != null) {
+            BLangVariable variableNode = (BLangVariable) onFailVarDefNode.getVariable();
+            if (variableNode.getKind() != NodeKind.VARIABLE && variableNode.getKind() != NodeKind.ERROR_VARIABLE) {
+                dlog.error(variableNode.pos, DiagnosticErrorCode.INVALID_BINDING_PATTERN_IN_ON_FAIL);
+            }
+
+            BType failErrorType;
+            LinkedHashSet<BType> onFailErrType = onFailErrTypes.peek();
+            boolean isVarTypeHardcoded = false;
+            if (onFailErrType.size() == 1) {
+                failErrorType = onFailErrType.iterator().next();
+            } else if (onFailErrType.size() > 1) {
+                failErrorType = BUnionType.create(null, onFailErrType);
+            } else {
+                failErrorType = symTable.errorType;
+                isVarTypeHardcoded = true;
+            }
+
             // Check on-fail node's variables and set types.
-            handleForeachDefinitionVariables(onFailVarDefNode, symTable.errorType,
-                    onFailClause.isDeclaredWithVar, true, onFailEnv);
+            handleForeachDefinitionVariables(onFailVarDefNode, failErrorType,
+                    onFailClause.isDeclaredWithVar, isVarTypeHardcoded, onFailEnv);
             BLangVariable onFailVarNode = (BLangVariable) onFailVarDefNode.getVariable();
-            if (!types.isAssignable(onFailVarNode.getBType(), symTable.errorType)) {
+
+            if (onFailVarNode.getBType() != null && !types.isAssignable(onFailVarNode.getBType(), symTable.errorType)) {
                 dlog.error(onFailVarNode.pos, DiagnosticErrorCode.INVALID_TYPE_DEFINITION_FOR_ERROR_VAR,
                         onFailVarNode.getBType());
             }
+
+            for (BType errorType : onFailErrTypes.peek()) {
+                if (onFailVarNode.getBType() != null && !types.isAssignable(errorType, onFailVarNode.getBType())) {
+                    dlog.error(onFailVarNode.pos, DiagnosticErrorCode.INCOMPATIBLE_ON_FAIL_ERROR_DEFINITION, errorType,
+                            onFailVarNode.getBType());
+                }
+            }
         }
+
+        onFailErrTypes.pop();
         data.env = onFailEnv;
         analyzeStmt(onFailClause.body, data);
     }
 
+    private void updateBreakMode(boolean onFailExists, BLangBlockStmt blockStmt, BLangOnFailClause onFailClause,
+                                 AnalyzerData data) {
+        if (onFailExists) {
+            blockStmt.failureBreakMode = getPossibleBreakMode(!data.commonAnalyzerData.errorTypes.peek().isEmpty());
+            this.analyzeNode(onFailClause, data);
```

Comment:
Can we rename this since it doesn't just update break mode, but also analyzes the node?